torch/_custom_op.py

```diff
@@ -226,43 +232,116 @@ def inner(f):
+            >>> def custom_nonzero_fake(x):
+            >>>     # Number of nonzero-elements is data-dependent
+            >>>     ctx = torch._custom_op.get_ctx()
+            >>>     nnz = ctx.new_data_dependent_symint()
+            >>>     # symbolic ints in PyTorch must be >= 2, so we constrain the
+            >>>     # range to at least 2. Note that the operator implementation
+            >>>     # must also do this.
+            >>>     ctx.constrain_range(nnz, min=2)
+            >>>     shape = [x.dim(), nnz]
+            >>>     result = x.new_empty(shape, dtype=torch.long)
+            >>>     return result
             >>>
-            >>> x = torch.randn(2, 3, device='meta')
-            >>> y = custom_sum(x, 1)
-            >>> assert y.shape == (2,)
+            >>> @numpy_nonzero.impl(['cpu', 'cuda'])
+            >>> def custom_nonzero_impl(x):
+            >>>     x_np = to_numpy(x)
+            >>>     res = np.stack(np.nonzero(x_np), axis=1)
+            >>>     # symbolic ints in PyTorch must be >= 2, so we constrain the
+            >>>     # range to at least 2.
+            >>>     if res.shape[0] <= 1:
+            >>>         raise RuntimeError("not supported")
+            >>>     return torch.tensor(res, device=x.device)
 
         """
 
         def inner(f):
-            library.impl(self._lib, self._opname, "Meta")(f)
+            frame = inspect.stack()[1]
+            if self._fake_impl is not None:
+                raise RuntimeError(
+                    f"Attempting to register a FakeTensor rule for operator {self._qualname} "
+                    f"that already has a FakeTensor rule registered from Python at "
+                    f"{self._fake_impl.location}. This is not supported."
+                )
+            new_location = f"{frame.filename}:{frame.lineno}"
+
+            # FakeTensor will look at _fake_impl
+            self._fake_impl = FuncAndLocation(f, new_location)
+
+            qualname = self._qualname
+
+            # Handle DispatchKey.Meta registration
+            @functools.wraps(f)
+            def f_with_ctx(*args, **kwargs):
+                def error_on_ctx():
+                    raise RuntimeError(
+                        f"Attempted to call get_ctx() for the meta implementation "
+                        f"for {qualname}."
+                        f"You have presumably called get_ctx() because the operator "
+                        f"has a data-dependent output shape; if so, there is no "
+                        f"such meta implementation and this error is the correct "
+                        f"behavior. Otherwise, please remove the call to get_ctx() "
+                        f"in the implementation registered with impl_fake "
+                        f"at {new_location}"
+                    )
+
+                with set_ctx_getter(error_on_ctx):
+                    return f(*args, **kwargs)
+
+            self._lib.impl(self._opname, f_with_ctx, "Meta")
```

Comment:
question to cement my understanding: Is it the case that this registration to the Meta key will basically never be called when using torch.compile, since we're basically guaranteeing that every custom op will have both a Meta impl and a FakeTensor impl, and FakeTensorMode always gives priority to the FakeTensor impl?

I guess the only case where we'd use this meta impl is if there's user code directly using meta tensors.